feat: New workflow action to send Whatsapp message

jemiluv8 opened the pull request:
## What does this PR do?
Add Support for the following whatsapp workflows

- send Whatsapp message to attendee
- send Whatsapp message to specific number

### Loom Demos
[Creating a whatsapp workflow by editing existing actions](https://www.loom.com/share/baa9de91fb40401d9d04d888d22c9a78)

[Creating whatsapp workflow action using the AddActionDialog](https://www.loom.com/share/2af273aaac7d4d9a821a0bdf6e82cdd1)

[Book meeting that has a whatsapp workflow to send whatsapp to attendee](https://www.loom.com/share/576844046e024772b6fc20dee9bce40c)

**Environment**: Staging(main branch) / Production
Fixes #3664
/claim #3664
## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration -->

- [ ] Get a sandbox whatsapp twilio number for the environment variable - TWILIO_WHATSAPP_PHONE_NUMBER
- [ ] Send a whatsapp message with the content `join liquid-rubber` to the twilio whatsapp number obtained
- [ ] Create a workflow to send whatsapp message to attendee or a specific number
- [ ] Verify that a whatsapp message is sent to your number 


## Checklist

<!-- Please remove all the irrelevant bullets to your PR -->
- I haven't added tests that prove my fix is effective or that my feature works

jemiluv8 committed:
add new Workflow actions for sending whatsapp message to attendee or a number

support creating new workflows with the whatsapp actions

jemiluv8 committed:
add cronjob to schedule scheduleWhatsappReminders
add WHATSAPP to WorkflowMethods with relevant migrations and schema updates

update twilioProvider methods to support sending whatsapp messages

introduce whatsappReminder manager module to support sending/scheduling/creating whatsapp reminders

update schedule reminder feature to work with WHATSAPP workflows

jemiluv8 committed:
modify AddActionDialog for workflow actions to require a sender when the workflow action is to WHATSAPP_ATTENDEE

jemiluv8 committed:
introduce TWILIO_WHATSAPP_PHONE_NUMBER in .env

support requiring attendee phone number if workflow action is WHATSAPP_ATTENDEE

Successfully triggered a whatsapp workflow on new booking.

TO FIX:
- text formatting of whatsapp message
- general review and testing of all possible events with whatsapp workflow

jemiluv8 committed:
introduce static templates for whatsapp actions
update scheduleWhatsappReminder to send properly formatted message

update default template for workflow to be based on trigger and type of action

jemiluv8 committed:
ensure whatsapp reminder is triggered when an event is cancelled and or rescheduled

refactored isSMSAction to separate it from isWhatsappAction

added isWhatsappOrSMSAction

jemiluv8 committed:
refactor workflow step container

jemiluv8 committed:
allow selecting template for whatsapp action

jemiluv8 committed:
Merge branch 'main' of https://github.com/jemiluv8/cal.com into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 committed:
update request reschedule and workflow update handler trpc handlers to schedule and delete whatsappMessages

jemiluv8 committed:
fix errors from upstream merge.

Whatsapp notifications for
- booking
- reschedule
- cancel

have been manually tested to work.

jemiluv8 committed:
modify number_sms_notifications to indicate SMS or Whatsapp notifications

jemiluv8 commented:
@CarinaWolli, I just created a draft PR to get some feedback on next steps.
There are a few things to “sort” out.

### WhatsApp Platform Issues

WhatsApp allows sending messages using only [approved templates](https://support.twilio.com/hc/en-us/articles/360007721954-Getting-Started-with-Twilio-for-WhatsApp). These templates can be managed via the Twilio account. This means we can’t have “CUSTOM” templates for WhatsApp workflow actions.

This limitation led me to create a couple of “default/proposed” templates for “WhatsApp” actions based on the event “triggers”. So there are templates for NEW_EVENT, EVENT_CANCELLED and the likes. Based on the platform limitations this made sense to me.

I expect we’d only need a handful of useful templates for sending things like meeting links.
#### I need assistance to decide on the templates we need and their content.

### WhatsApp Templates and UX

- I don’t allow editing any of the WhatsApp templates. This is supposed to be readonly since we can only send WhatsApp messages using “approved” templates.

jemiluv8 committed:
Merge branch 'main' of https://github.com/jemiluv8/cal.com into NEW-WHATSAPP-WORKFLOW-ACTIONS

CarinaWolli commented:
@jemiluv8 awesome work so far 🙌🏻

> I need assistance to decide on the templates we need and their content. 

@Jaibles could you take a look at that? Here are the templates that got defined in that PR:
<img width="839" alt="Screenshot 2023-05-10 at 14 56 39" src="https://github.com/calcom/cal.com/assets/30310907/ec268f5a-bee8-44d7-a87a-ba6d17de29a1">
<img width="809" alt="Screenshot 2023-05-10 at 14 56 43" src="https://github.com/calcom/cal.com/assets/30310907/c467f2ce-2aeb-4180-bae7-806a7f3a5609">
<img width="827" alt="Screenshot 2023-05-10 at 14 56 47" src="https://github.com/calcom/cal.com/assets/30310907/a89866c9-7e7c-4193-b498-1c0c5094b424">

jemiluv8 committed:
minor refactoring

jemiluv8 commented:
@Jaibles is there any update on the content for the Whatsapp templates?

Jaibles commented:
> @Jaibles is there any update on the content for the Whatsapp templates?

Sorry, I missed this. I will take a look now.

Jaibles commented:
@jemiluv8 Can we add any formatting? Some bolds for the event titles would be good. 

Some small changes

**Cancelled**
Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on
{EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} {TIMEZONE} has been cancelled.

**Completed**
Hi {ATTENDEE}, thank you for attending the event ({EVENT_NAME}) on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} {TIMEZONE}.

**Rescheduled**
Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on
{EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} {TIMEZONE} has been rescheduled.

jemiluv8 commented:
> @jemiluv8 Can we add any formatting? Some bolds for the event titles would be good.
> 
> Some small changes
> 
> **Cancelled** Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} ({TIMEZONE}) has been cancelled.
> 
> **Completed** Hi {ATTENDEE}, thank you for attending the event ({EVENT_NAME}) on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} ({TIMEZONE}).
> 
> **Rescheduled** Hi {ATTENDEE}, your meeting ({EVENT_NAME}) with {ORGANIZER} on {EVENT_DATE_YYYY MMM D} at {START_TIME_h:mmA} ({TIMEZONE}) has been rescheduled.

@Jaibles that is great. I'll bolden the event titles.

jemiluv8 committed:
update whatsapp templates as per design and run some manual tests

jemiluv8 committed:
fix bugs and test scheduleWhatsappReminders handler for the CRON JOB

jemiluv8 committed:
Merge branch 'main' into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 committed:
fix linting issues and bolden even names in whatsapp templates

jemiluv8 committed:
modify getBookingResponsesSchema
- applied duct-tape to phone schema parsing - for whatever reasons, the plus in phone number input is strimmed out before it is sent to the zod schema validator

Applied yet more duct tape to twilio provider to deal with the missing plus sign again. Not sure this "lazy" application of duct-tape is the way to go but it works and will be refactored. We have to rethink this

update activateEventType.handler to schedule whatsapp reminder

jemiluv8 commented on packages/features/bookings/lib/getBookingResponsesSchema.ts:
```diff
@@ -177,7 +177,8 @@ function preprocess<T extends z.ZodType>({
         if (bookingField.type === "phone") {
-          if (!phoneSchema.safeParse(value).success) {
+          const phoneNumber = value && value.startsWith("+") ? value : `+${value}`
+          if (!phoneSchema.safeParse(phoneNumber).success) {
```
@CarinaWolli,

I had the issue in the screenshot below and ended up having to "patch" this file and the zod schema validation.
I don't like my patch and have been investigating the source of the missing "+" from phone number inputs.

A recent commit however changed the react library used for our phone input from "react-phone-number-input" 
to "react-phone-input-2". I'm not sure what to make of this yet but prior to that the "+" prefix was added to the
value of the phone number input.

I'm reaching out to decide what to do about this. I wonder if bookings that require entering phone number are going through (assuming the recent changes have been deployed)

<img width="854" alt="Screenshot 2023-06-01 at 12 25 53 PM" src="https://github.com/calcom/cal.com/assets/119384208/c73c3996-2ffa-4a35-86eb-5dc4c8a38dba">

CarinaWolli replied:
This has been fixed already here: https://github.com/calcom/cal.com/pull/9258

Sorry for that!

jemilu2 replied:
Thanks.

jemiluv8 committed:
revert changes to getBookingResponses
modify getSenderId to return different numbers for whatsapp

jemiluv8 committed:
Merge branch 'main' of https://github.com/jemiluv8/cal.com into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 committed:
fix issue when saving workflow that sends a whatsapp message to a specific number

add NEXT_PUBLIC_SENDER_ID_WHATSAPP

PeerRich commented:
thank you for marking as "for review". we will try to review this this week, there are some other higher priority PRs right now tho which have to be reviewed first: https://github.com/calcom/cal.com/pulls?q=is:pr+is:open+sort:updated-desc+label:%22High+priority%22

CarinaWolli commented on packages/features/ee/workflows/lib/actionHelperFunctions.ts:
```diff
@@ -1,9 +1,74 @@
+    default:
+      return whatsappReminderTemplate
+  }
+}
```
new line missing

CarinaWolli commented on packages/features/ee/workflows/lib/constants.ts:
```diff
@@ -14,11 +14,32 @@ export const WORKFLOW_ACTIONS = [
+  WorkflowTemplates.REMINDER,
+] as const;
+
+export const WHATSAPP_WORKFLOW_TEMPLATES = [
```
We can also add these new templates for SMS (but we can do that in a follow-up PR)

jemiluv8 replied:
I'll be happy to do it in a follow-up PR. This one feels too big already.

CarinaWolli commented on packages/features/ee/workflows/lib/reminders/templates/whatsapp/index.ts:
```diff
@@ -0,0 +1,4 @@
+export * from "./whatsappEventCancelledTemplate"
+export * from "./whatsappEventCompletedTemplate"
+export * from "./whatsappEventReminderTemplate"
+export * from "./whatsappEventRescheduledTemplate"
```
new line missing

CarinaWolli commented on packages/prisma/migrations/20230420222915_add_whatsapp_workflow_actions/migration.sql:
```diff
@@ -0,0 +1,10 @@
+-- With PostgreSQL versions 11 and earlier, this is not possible
+-- in a single migration. This can be worked around by creating
+-- multiple migrations, each migration adding only one value to
+-- the enum.
```
@zomars are we supporting PostresSQL version 11 and earlier? If yes, we would need a separate migration file for each enum value added

CarinaWolli replied:
Should be fine to have them in one migration file 👍
Could you add all migrations into one single migration file?

jemiluv8 replied:
I'll do that.

CarinaWolli reviewed (commented):
Great work on what I have seen so far 👏🏻  I left you some first feedback, but I didn't find the time yet to review and test in detail. I also started to look into what needs to be done to make it work in production and submitted one first template to Twilio which is waiting for approval currently :) 

Maybe you can already address my comments, and I will come back soon to finish reviewing

CarinaWolli commented on packages/features/ee/workflows/components/WorkflowStepContainer.tsx:
```diff
@@ -727,6 +770,7 @@ export default function WorkflowStepContainer(props: WorkflowStepProps) {
                       updateTemplate={updateTemplate}
                       firstRender={firstRender}
                       setFirstRender={setFirstRender}
+                      editable={!isWhatsappAction(step.action)}
```
type error
![Screenshot 2023-06-15 at 10 44 43](https://github.com/calcom/cal.com/assets/30310907/099d3a24-45cf-446a-8341-236cb91a3219)

jemiluv8 commented on packages/features/ee/workflows/lib/reminders/smsProviders/twilioProvider.ts:
```diff
@@ -19,27 +19,39 @@ function assertTwilio(twilio: TwilioClient.Twilio | undefined): asserts twilio i
+}
+
+function getSMSNumber(phone: string, whatsapp = false) {
+  return whatsapp ? `whatsapp:${phone}` : phone;
```
No I don't know something different. I was actually using getDefaultSender for whatsapp initially. Will change it back.

jemiluv8 replied:
I have thus far relied on the implementation of SMS workflow as a guide. And when sending to attendee the SMS workflow shows a SENDER id similar to what I do for Whatsapp workflow(in the screenshot below).

As far as I was concerned any valid twilio whatsapp number can be used as a Sender Id (just like it is for SMS). The current implementation uses the Sender ID when  specified. The change you suggested will just ignore it.

To be consistent with everything else, I will also need to remove the Sender ID input box in the screenshot below for whatsapp related workflows. 

Should I proceed with both changes?

<img width="743" alt="Screenshot 2023-06-15 at 9 55 56 AM" src="https://github.com/calcom/cal.com/assets/119384208/3bea7bd2-a729-4869-8a7b-41605d61d242">

CarinaWolli replied:
Yes, please make both of the changes. Thank you 🙏

jemiluv8 replied:
I just made both changes.

jemiluv8 committed:
update whatsapp templates
- set character limit at 1024
- avoid boldening the brackets around event name

remove duplicate editable prop.

merge values for duplicate editable prop on the editor

jemiluv8 committed:
use twilio whatsapp number as sender when scheduling whatsapp notifications and remove alphaNumeric sender for whatsapp

CarinaWolli committed:
Merge branch 'main' into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 committed:
remove redundant sender argument from scheduleWhatsappReminder - we use the twilio whatsapp number instead for all

jemiluv8 committed:
Merge branch 'NEW-WHATSAPP-WORKFLOW-ACTIONS' of https://github.com/jemiluv8/cal.com into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 committed:
refactored getWhatsappTemplateForAction to reuse getWhatsappTemplateFunction

refactor actionHelperFunctions.ts

CarinaWolli commented:
I am currently working on getting the WhatsApp templates approved in Twilio. I created the reminder template at first and it got approved right away. But canceled and rescheduled template got rejected, without any reason given. I resubmitted it and will see if it goes through this time

CarinaWolli reviewed (changes requested):
I tested it and finished my review. It all looks good to me, I just left you some minor comments. Please address them and I will do my best to get the Twilio templates approved, so we can merge it. Thank you 🙏

jemiluv8 committed:
Made modifications - mostly refactored stuff - based on review.

jemiluv8 commented on packages/features/ee/workflows/lib/reminders/templates/whatsapp/whatsappEventCancelledTemplate.ts:
```diff
@@ -0,0 +1,35 @@
+
+  const templateOne = `Hi${
+    name ? ` ${name}` : ``
+  }, your meeting (*${eventName}*) with ${attendee} on ${eventDate} at ${startTime} ${timeZone} has been cancelled.`
```
resolved

jemiluv8 commented on packages/features/ee/workflows/pages/workflow.tsx:
```diff
@@ -221,7 +221,7 @@ function WorkflowPage() {
           //check if phone number is verified
           if (
-            step.action === WorkflowActions.SMS_NUMBER &&
+            (step.action === WorkflowActions.SMS_NUMBER  || step.action === WorkflowActions.WHATSAPP_NUMBER)&&
```
resolved.

jemiluv8 commented on packages/features/ee/workflows/components/WorkflowDetailsPage.tsx:
```diff
@@ -98,10 +98,10 @@ export default function WorkflowDetailsPage(props: Props) {
       numberRequired: numberRequired || false,
-      sender: isSMSAction(action) ? sender || SENDER_ID : SENDER_ID,
-      senderName: !isSMSAction(action) ? senderName || SENDER_NAME : SENDER_NAME,
+      sender: isSMSOrWhatsappAction(action) ? sender || SENDER_ID : SENDER_ID,
```
Reverted to original. Resolved.

jemiluv8 commented on packages/features/ee/workflows/components/AddActionDialog.tsx:
```diff
@@ -102,6 +102,15 @@ export const AddActionDialog = (props: IAddActionDialog) => {
+      } else if (newValue.value === WorkflowActions.WHATSAPP_ATTENDEE) {
+        setIsSenderIdNeeded(false);
+        setIsEmailAddressNeeded(false);
+        setIsPhoneNumberNeeded(false);
```
resolved

jemiluv8 committed:
Removed no longer used NEXT_PUBLIC_SENDER_ID_WHATSAPP from .env.example

jemiluv8 committed:
reduce prisma migration files for

jemiluv8 committed:
Merge branch 'main' of https://github.com/jemiluv8/cal.com into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 commented on packages/features/ee/workflows/pages/workflow.tsx:
```diff
@@ -142,7 +142,7 @@ function WorkflowPage() {
           ...step,
           senderName: step.sender,
-          sender: isSMSAction(step.action) ? step.sender : SENDER_ID,
+          sender: isSMSOrWhatsappAction(step.action) ? step.sender : SENDER_ID,
```
resolved

jemiluv8 commented on apps/web/public/static/locales/en/common.json:
```diff
@@ -1465,7 +1469,7 @@
   "team_url_taken": "This URL is already taken",
   "team_publish": "Publish team",
-  "number_sms_notifications": "Phone number (SMS notifications)",
+  "number_sms_notifications": "Phone number (SMS or Whatsapp notifications)",
```
Not sure about this.

There can be SMS and WHATSAPP actions associated with the same event type so
the client may end up receiving both SMS or WHATSAPP actions (for the same event)
depending on the workflows associated with the event type. 

This means it may not just be a case of SMS or Whatsapp it could be
1. SMS
2. Whatsapp
3. SMS and WHATSAPP

jemiluv8 replied:
Kindly check [this](https://github.com/calcom/cal.com/blob/931763959b288776514e0b90597fe18d2dc7cf08/packages/features/bookings/lib/getBookingFields.ts#L22) and related code for context and let me know how we proceed.

CarinaWolli replied:
I think to keep it simple we can just call it "Text notifications" instead

jemiluv8 replied:
I've updated this to show "Phone Number (Text Notifications)".
Changed the related locale variable to number_text_notifications

jemiluv8 committed:
fix issue from review

jemiluv8 committed:
change locale variable "number_sms_notifications" to "number_text_notifications" to reflect the fact that the notification may come via multiple "text" channels - sms or whatsapp.

Updated variable names for en/common.js and fr/common.json - The french translation remains the same according to google translate

jemiluv8 committed:
Merge branch 'main' of https://github.com/jemiluv8/cal.com into NEW-WHATSAPP-WORKFLOW-ACTIONS

jemiluv8 commented on .env.example:
```diff
@@ -110,9 +110,11 @@ TWILIO_SID=
 # For NEXT_PUBLIC_SENDER_ID only letters, numbers and spaces are allowed (max. 11 characters)
 NEXT_PUBLIC_SENDER_ID=
 TWILIO_VERIFY_SID=
+NEXT_PUBLIC_SENDER_ID_WHATSAPP=
```
removed

CarinaWolli reviewed (changes requested):
Found some more issues: 
1) I saved a phone number into my workflow, when I revisit the workflow I don't see the number anymore but it says still verified. This looks like a new issue to me:
<img width="834" alt="Screenshot 2023-07-03 at 15 56 22" src="https://github.com/calcom/cal.com/assets/30310907/e93613b4-4279-41d0-81c9-bf5248f68622">

2) I am able to choose WhatsApp templates for non-WhatsApp actions: 
- choose a WhatsApp action 
- choose completed template
- change action to email action 
- see that template is still completed (we should fall back to custom)
- save workflow with the completed template

Again a big thank you for your work so far, really love that PR 🙏 Just tested the reminder and completed template outside of sandbox mode and everything worked 🙌🏻 (We maybe have to disable the other two templates for now until I get them approved in Twilio)

jemiluv8 committed:
Update packages/features/ee/workflows/lib/reminders/whatsappReminderManager.ts

set reasonable default for name in whatsappReminder template -

Co-authored-by: Carina Wollendorfer <30310907+CarinaWolli@users.noreply.github.com>

jemiluv8 committed:
prevent user from setting invalid template a particular workflow action

CarinaWolli commented:
> I've tried the same test in chrome and firefox. I was able to replicate it in chrome but not firefox.
> So it works as expected in firefox. Can you confirm this?

Yes can confirm, it works in Firefox not in Chrome

> Should I go ahead and disable the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.

Yes, let's disable it for now. We should still keep all the migrations and everything, let's just not show them in the template dropdown. We want it to be simple to enable them once Twilo approves them

jemiluv8 committed:
comment out cancelled and rescheduled templates until they are approved

jemiluv8 commented:
I've disabled the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.

CarinaWolli commented:
> I've disabled the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.

@jemiluv8  Perfect! Any ideas already why the phone number isn't showing in Chrome?

CarinaWolli commented on packages/features/ee/workflows/lib/constants.ts:
```diff
@@ -14,11 +14,33 @@ export const WORKFLOW_ACTIONS = [
+export const WORKFLOW_TEMPLATES = [
+  WorkflowTemplates.CUSTOM,
+  WorkflowTemplates.REMINDER,
+  WorkflowTemplates.REMINDER,
```
```suggestion
```

jemiluv8 replied:
I've reverted the changes to remove the "CANCELLED" and "RESCHEDULED" templates

CarinaWolli replied:
Can you remove the duplicate `WorkflowTemplates.REMINDER` in `WORKFLOW_TEMPLATES` as well please

jemiluv8 replied:
I just did

CarinaWolli commented:
@jemiluv8 Good news, the templates just got approved 🎉

I will test them in a little bit, but we can revert the changes now :)

jemiluv8 committed:
revert changes to WHATSAPP_WORKFLOW_TEMPLATES constant.

Restore cancelled and resheduled templates

jemiluv8 committed:
remove duplicate reminder template name in WORKFLOW_TEMPLATES

jemiluv8 commented:
> > I've disabled the "CANCELLED" and "RESCHEDULED" templates for the Whatsapp workflow actions.
> 
> @jemiluv8 Perfect! Any ideas already why the phone number isn't showing in Chrome?

@CarinaWolli 
At the moment no. But I'm of the view that it is a general issue that may have to do with the
react-phone-input-2 introduced not long ago. I'd recommend we wait for a fix before "potentially"
merging this PR.

jemiluv8 commented:
@CarinaWolli I'm still looking into the phone input not showing in chrome.
During my experiments I've noticed that I sometimes got the number to 
show while editing code (after hot reload).

Sometimes the number shoes and disappears - very fast. I should be able to get
a fix soon.

jemiluv8 commented:
> @jemiluv8 Let me know in case you need help for the phone number issue, I think that's the last thing missing now 🙏

@CarinaWolli I do need help. But perhaps starting tomorrow if I'm not able to come up with a solution.
I've been trying stuff without a lot of progress but I haven't yet exhausted all my possible attempts at a solution.

jemiluv8 committed:
Modify phone input to handle a special case of not getting a valid response from the trpc endpoint for country code.

This occurs on local development when 'x-vercel-ip-country' header is not properly set because the app is running locally

The response to the country code query returns an empty string. In this case perhaps we shouldn't update the country set on the phone input?

jemiluv8 commented:
@CarinaWolli, I just pushed a hot fix for the phone number thing.
Can you confirm it works in chrome as well.

jemiluv8 committed:
fix bug with verified phone number showing as unverified when switching between sms and whatsapp actions that require a number

Commit:
remove french translation (needs to be updated as well)

Commit:
don't set reminders for whatsapp_attendee if booking was already made + clean up

CarinaWolli committed:
Merge branch 'main' into NEW-WHATSAPP-WORKFLOW-ACTIONS

CarinaWolli reviewed (approved):
I made some minor final adjustments, and now we are good to go 🙌🏻
Thank you for your excellent work 🙏
